contracts/contracts/Staking.sol

```diff
@@ -0,0 +1,225 @@
+
+    // -------------------------------------------------------------------------
+
+    struct PendingDelegation {
+        address from;
+        StakingAddress to;
+        uint128 amount;
+    }
+
+    struct PendingUndelegation {
+        StakingAddress from;
+        address payable to;
+        uint128 shares;
+        uint64 endReceiptId;
+        uint64 epoch;
+    }
+
+    struct UndelegationPool {
+        uint128 totalShares;
+        uint128 totalAmount;
+    }
+
+    // -------------------------------------------------------------------------
+
+    event OnDelegateStart(
+        address indexed who,
+        StakingAddress to,
+        uint256 amount,
+        uint64 receiptId
+    );
+
+    event OnDelegateDone(uint64 indexed receiptId, address who, uint128 shares);
+
+    event OnUndelegateStart(
+        uint64 indexed receiptId,
+        address who,
+        uint64 epoch,
+        uint128 shares
+    );
+
+    // -------------------------------------------------------------------------
+
+    constructor() {
+        // Due to an oddity in the oasis-cbor package, we start at 2**32
+        // Otherwise uint64 parsing will fail and the message is rejected
+        lastReceiptId = 4294967296;
+    }
+
+    /**
+     * Begin or increase delegation by sending an amount of ROSE to the contract.
+     *
+     * Delegation will fail if the minimum per-validator amount has not been
+     * reached, at the time of writing this is 100 ROSE.
+     *
+     * Only one delegation can occur per transaction.
+     *
+     * @param to Staking address of validator on the consensus level
+     */
+    function delegate(StakingAddress to) public payable returns (uint64) {
+        // Whatever is sent to the contract is delegated.
+        require(msg.value < type(uint128).max);
+
+        uint128 amount = uint128(msg.value);
+
+        uint64 receiptId = lastReceiptId++;
+
+        Subcall.consensusDelegate(to, amount, receiptId);
+
+        pendingDelegations[receiptId] = PendingDelegation(
+            msg.sender,
+            to,
+            amount
+        );
+
+        emit OnDelegateStart(msg.sender, to, msg.value, receiptId);
+
+        return receiptId;
+    }
+
+    /**
+     * Retrieve the number of shares received in return for delegation.
+     *
+     * The receipt will only be available after the delegate transaction has
+     * been included in a block. It is necessary to wait for the message to
+     * reach the consensus layer and be processed to determine the number of
+     * shares.
+     *
+     * @param receiptId Receipt ID previously emitted/returned by `delegate`.
+     */
+    function delegateDone(uint64 receiptId) public returns (uint128 shares) {
+        PendingDelegation memory pending = pendingDelegations[receiptId];
+
+        require(pending.from != address(0), "unknown receipt");
+
+        shares = Subcall.consensusTakeReceiptDelegate(receiptId);
+
```

Comment:
That's correct.

There is also a bunch of other accounting stuff I need to implement which, after starting an initial implementation of a staking contract for a dApp, it would need to be fully functional.

I'm going to move Staking.sol into a different project as I don't think it belongs in the sapphire-contracts repo, but the Subcall.sol does.